llvm/include/llvm/IR/PassManager.h

```diff
@@ -64,6 +64,8 @@ extern llvm::cl::opt<bool> UseNewDbgInfoFormat;
 
 namespace llvm {
 
+class MachineFunction;
```

Comment:
We shouldn't have any of the MIR machinery in the files meant for IR. My hesitation is MIR being a target-specific entity, an edge case for a specific target might require some of its properties to be exposed here. I understand we want to have a unified PM for both IR and MIR passes in the LLC pipeline. But can't we have a separate folder to accommodate designs common to IR and MIR infrastructure?

Reply:
I wasn't talking about this instance alone. But in general, while designing the machinery for NPM backend passes.

Reply:
I agree it would be ideal to move any mention of `MachineFunction` out of `llvm/IR`, but this functionality is tricky to support in a clean way

we'd like for a machine function pass to be able to say what invariants the MIR holds (via `MachineFunctionProperties`) and for the pass manager infra to verify that. we could always tell every pass to verify these properties before and after running. initially I thought this would be a bunch of extra boilerplate, but actually looking at `MachineFunctionPass::get[Required,Set,Cleared]Properties`, maybe this is more feasible than I initially imagined. `Set`/`Cleared` aren't very common, so those passes could just manually set/clear the properties on `MF.getProperties()`. as for `Required`, currently any legacy pass that wants to ensure the incoming `MachineFunction` has certain properties needs to override `getRequiredProperties()`. it might actually be less overall code if we had a helper function to verify properties that passes would call, e.g.

```
diff --git a/llvm/include/llvm/CodeGen/MachinePassManager.h b/llvm/include/llvm/CodeGen/MachinePassManager.h
index 7713c55661cc..669076e47634 100644
--- a/llvm/include/llvm/CodeGen/MachinePassManager.h
+++ b/llvm/include/llvm/CodeGen/MachinePassManager.h
@@ -44,7 +44,20 @@ using MachineFunctionAnalysisManager = AnalysisManager<MachineFunction>;
 /// automatically mixes in \c PassInfoMixin.
 template <typename DerivedT>
 struct MachinePassInfoMixin : public PassInfoMixin<DerivedT> {
-  // TODO: Add MachineFunctionProperties support.
+  void verifyMachineFunctionProperties(MachineFunction&MF, MachineFunctionProperties RequiredProperties) {
+#ifndef NDEBUG
+  if (!MF.getProperties().verifyRequiredProperties(RequiredProperties)) {
+    errs() << "MachineFunctionProperties required by " << DerivedT::name()
+           << " pass are not met by function " << MF.getName() << ".\n"
+           << "Required properties: ";
+    RequiredProperties.print(errs());
+    errs() << "\nCurrent properties: ";
+    MF.getProperties().print(errs());
+    errs() << "\n";
+    llvm_unreachable("MachineFunctionProperties check failed");
+  }
+#endif
+  }
 };
 ```

and then a pass that wants to verify would do so like

```
FooPass::run(MachineFunction&MF, ...) {
  verifyMachineFunctionProperties(MF, MachineFunctionProperties().set(
        MachineFunctionProperties::Property::NoVRegs));
}
```

@arsenm @paperchalice thoughts on this approach?

as for other approaches:

initially I was hoping we might be able to augment the return value (`PreservedAnalyses`) of the pass here (in fact the return value was template parameterized at some point), but we also need to verify before the pass runs, so that doesn't work

I tried templatizing `PassConceptT`, meaning we'd specify it in the `MachineFunctionPassManager` declaration, but it turned out to be harder than I thought for reasons I can't remember now

Reply:
either way the pass author needs to write a couple of lines of code to say to check certain properties; IMO putting it in the actual pass makes it more explicit and less magicky

is there a reason you say that it would be more annoying to put in the actual pass? I think adding pass manager infra for this complicates things a lot

Reply:
Mostly the API is bad. We have this full set of virtual functions to implement, and then you have to construct an empty MachineFunctionProperties, and set fields in it. It would be much more convenient to have something declarative

Reply:
> It would be much more convenient to have something declarative

I still don't really understand this.

```
FooPass::run(MachineFunction&MF, ...) {
  verifyMachineFunctionProperties(MF, MachineFunctionProperties().set(
        MachineFunctionProperties::Property::NoVRegs));
}
```

is less code and more explicit than writing an extra method that also returns the same `MachineFunctionProperties`. Can you give an example of what you have in mind?

> and then you have to construct an empty MachineFunctionProperties, and set fields in it

`MachineFunctionProperties::set()` is like a builder, so constructing an arbitrary `MachineFunctionProperties` can be expressed in one statement

> I would prefer verify machine function properties in instrumentation (e.g. by a string map)

can you clarify what you mean by "by a string map"?

> modify properties in pass, e.g. by a Scope-Bound Resource Management object

I agree it's nice to be able to say at the beginning of `run` that some MachineFunction property will be set when the pass returns since putting the `verify` and `set` `MachineFunctionProperties` together is nice, but also I don't think it's a huge deal to have the pass manually set `MachineFunctionProperties` at the end of the pass. And IMO the boilerplate you're proposing is not worth it

Reply:
> FooPass::run(MachineFunction&MF, ...) {
>   verifyMachineFunctionProperties(MF, MachineFunctionProperties().set(
>         MachineFunctionProperties::Property::NoVRegs));
> }
> ```

This guarantees future bugs from either forgetting to add this, or being clever with putting code before this.


> `MachineFunctionProperties::set()` is like a builder, so constructing an arbitrary `MachineFunctionProperties` can be expressed in one statement

A builder here is obnoxious. These constraints are expressible as a constant bitfield, the API shouldn't be built around mutation of an object